AGENT-997: Internal dev docs for authentication-authorization (#9224)
openshift/installer: docs/user/agent/authentication-authorization.md

# Agent-Based Installer Authentication and Authorization - OpenShift 4.18

## Introduction
In OpenShift 4.18, the Assisted Service APIs used by the Agent-Based Installer are now secured through authentication and authorization mechanisms. This added security ensures that only authorized services can interact with the installer and other system components. This document provides an overview of how to enable authentication and authorization for the agent-based installer.

## Enabling Authentication and Authorization

To enable authentication and authorization, the `AuthType` setting must be configured in the `assisted-service.env.template` file. Set the following value:

```bash
AuthType = "agent-installer-local"
```

This setting ensures that the authentication mechanism required by the agent-based installer is activated.

## Tokens for Agent-Based Installer

> When creating an Agent ISO or PXE files on Day 1, the OpenShift installer generates three separate authentication tokens: `agentAuth`, `userAuth`, and `watcherAuth`. These tokens serve different purposes and have varying levels of access.

Reviewer: Please also describe where the tokens are generated (ie how they are stored in the ISO)

> When creating an Agent ISO or PXE files on Day 1, the OpenShift installer generates three separate authentication tokens: `agentAuth`, `userAuth`, and `watcherAuth`. These tokens serve different purposes and have varying levels of access.

Author: Done

### Types of Tokens:
1. **agentAuth**:

>    - **Access Level**: Read-Write (RW)

Author: Added

>    - **Access Level**: Read-Write (RW)

Author: Please see below section - `Managing Access with Tokens and Security Schemes in swagger.yaml`

   - **Used By**: The agent service that runs on hosts booted with the agent-based installer.
   
2. **userAuth**:
   - **Access Level**: Read-Write (RW)

>    - **Used By**: Systemd services and the agent-based installer client in the Assisted Service.

Author: Done

   
3. **watcherAuth**:
   - **Access Level**: Read-Only (RO)

>    - **Used By**: Only the `wait-for` command (on Day 1) and `monitor-add-nodes` command (on Day 2).

Reviewer: ```suggestion
   - **Used By**: The monitoring commands (the `wait-for` command on Day 1 and `monitor-add-nodes` command on Day 2).
```

### Token Expiry Behavior

> - **Day 1** Tokens: The tokens generated on Day 1 do not expire, meaning the agent ISO remains valid indefinitely. This is useful for use cases where customers create an agent ISO on one day and boot it on a later day (e.g., in a data center).

Reviewer: ```suggestion
- **Day 1** Tokens: The tokens generated on Day 1 never expire, meaning the agent ISO remains valid indefinitely. This is useful for use cases where customers create an agent ISO on one day and boot it on a later day (e.g., in a data center).
```

> - **Day 2** Tokens: Tokens generated for nodes ISO on Day 2 expire after 48 hours. This ensures that the nodes.iso is only valid for a short period.

Reviewer: ```suggestion
- **Day 2** Tokens: Tokens generated for nodes ISO on Day 2 expire after 48 hours. This ensures that the ISO is only valid for a short period.
```

> ### Token Management

Reviewer: ```suggestion
### Day 2 token management
```

When a Day 2 node ISO is created, all three tokens are saved in the `openshift-config` namespace as a new secret named `agent-auth-token`. The following steps outline how token management works:

> 1. If the secret exists in the cluster, the token can be retrieved.

Reviewer: ```suggestion
1. If the `openshift-config/agent-auth-token` secret already exists in the cluster, the tokens are retrieved from the secret (instead of being generated).
```

Reviewer: Also, what happens if does not exist?

Author: Added

> 2. All tokens expire at the same time, so only one token needs to be checked for expiry.

Reviewer: This does not seem specific to day 2 only, and a very internal detail, I'd remove it from here (or at least move it into another section)

> 3. The expiration time is calculated as 24 hours before the token's expiry time.

Reviewer: Can you please elaborate it better? It does not seem extremely clear to me here (I'd remove it at all, the next two points explain the process better IMHO)

4. If the token in the secret is older than 24 hours when creating a node ISO, the installer generates new tokens. These new tokens are saved in the asset store, and the secret in the cluster is updated with the new token values.

> 5. If the token in the secret is not older than 24 hours, the installer retrieves the existing token from the cluster, updates the asset store with the retrieved token, and uses the retrieved token for the new node ISO.

Reviewer: This looks like something that could be also shown better using a simple diagram. Have you maybe ever try [mermaid](https://github.com/mermaid-js/mermaid) for that?

Author: Sure, I will add it.

> ## Authorization

Reviewer: This section seems more specific to AS

> To implement authorization for the Agent-Based Installer, a new security definition is added in the `swagger.yaml` file in Assisted Service. The new security definition is as follows:

Reviewer: ```suggestion
To implement authorization for ABI, a new security definition is added in the `swagger.yaml` file in Assisted Service. The new security definition is as follows:
```

```yaml
watcherAuth:
  type: apiKey
  in: header
  name: Watcher-Authorization
```

The required endpoints are annotated with this security definition. For example:

```yaml
/v2/clusters:
  get:
    tags:
      - installer
    security:
      - userAuth: [admin, read-only-admin, user]
      - watcherAuth: []
    description: Retrieves the list of OpenShift clusters.
    operationId: v2ListClusters
```

### Authorization Handler

> The Assisted Service API includes an authorization handler implemented in the `pkg/auth/agent_local_authz_handler.go` file. The `AgentLocalAuthzHandler`  is responsible for verifying that the `auth_scheme` in the token matches the required `authScheme` for the endpoint. The authorization logic works as follows:

Reviewer: A direct link to the real code could be useful here

Author: Done

## Authentication API (Server Side)

The authentication server validates the token provided by the client using the `validateToken` function. The following steps explain the server-side validation process:

1. When an agent or nodes ISO is created, a set of public and private keys is generated. The private key is used to sign the JWT tokens, while the associated public key is made available to the API server.
2. The token is passed to the API server as part of the request. The API server validates the token by verifying its signature against the public key. This ensures the token was signed by the correct private key and has not been tampered with.
3. If the token is invalid or expired, an error is logged and the request is rejected with a `Unauthorized` error.
4. The claims in the token are extracted.
5. If the `exp` (expiration) claim is found, the server checks whether the token has expired.
   - In the install workflow, the `exp` claim will not be present.
   - In the add-nodes workflow, the token expiration is checked. If expired, a new token must be generated.
6. If the token is valid and not expired, the claims are returned and the request proceeds.

> ## Expired Token Handling on Day 2

Reviewer: This entire paragraph looks like it could be integrated in the previous token management one, where the behavior for day 2 have been described

When a node is booted using the Day 2 `nodes.iso` to add worker nodes, the `agent-auth-token-status.service` systemd service will check if the authentication tokens have expired. If the tokens are expired, the following message will be displayed:

```
The authentication token has expired. Please generate a new ISO using the "oc adm node-image create" command, then reboot the node.
```

This ensures that the user is notified of expired tokens and can take appropriate action to generate a new ISO.

> ## Conclusion

Reviewer: Is that really needed? It seems to me that the motivations have been clearly described in the introduction paragraph

# Agent-Based Installer Authentication and Authorization - OpenShift 4.18

## Introduction

> Since OpenShift 4.18, the Assisted Service (AS) APIs used by the Agent-Based Installer (ABI) have been secured through authentication and authorization mechanisms. This added security ensures that only authorized services can interact with AS. This document provides an overview of how the authentication and authorization mechanisms implemented for ABI.

Reviewer 2: ```suggestion
Since OpenShift 4.18, the Assisted Service (AS) APIs used by the Agent-Based Installer (ABI) have been secured through authentication and authorization mechanisms. This added security ensures that only authorized services can interact with AS. This document provides an overview of how the authentication and authorization mechanisms were implemented for ABI.
```

## Enabling Authentication and Authorization

To enable authentication and authorization, the `AuthType` setting must be configured in the `assisted-service.env.template` file. Set the following value:

```bash
AuthType = "agent-installer-local"
```

This setting ensures that the authentication mechanism required by the agent-based installer is activated and is an internal configuration.

## Tokens for Agent-Based Installer

> When creating an Agent ISO (or PXE artifacts) for installing a new cluster, ABI generates three separate authentication tokens: `agentAuth`, `userAuth`, and `watcherAuth`. These tokens serve different purposes and have varying levels of access. The tokens are the part of the ignition that gets embeded in the CoreOS base ISO along with other ABI specific configurations for a cluster. The tokens are stored `/usr/local/share/assisted-service/assisted-service.env` in the ISO. The base64 encoded public key required to validate these tokens is also saved in the same file. Additionally, this data also gets stored in a hidden file `.openshift_state_install.json` present in the directory location mentioned when creating ISO.

Reviewer 2: ```suggestion
When creating an Agent ISO (or PXE artifacts) for installing a new cluster, ABI generates three separate authentication tokens: `agentAuth`, `userAuth`, and `watcherAuth`. These tokens serve different purposes and have varying levels of access. The tokens are the part of the ignition that gets embeded in the CoreOS base ISO along with other ABI specific configurations for a cluster. The tokens are stored in `/usr/local/share/assisted-service/assisted-service.env` in the ISO. The base64 encoded public key required to validate these tokens is also saved in the same file. Additionally, this data also gets stored in a hidden file `.openshift_state_install.json` present in the directory location mentioned when creating ISO.
```
Why is it stored in the hidden file?

Author: The tokens are part of the ignition and is also an asset hence those are saved in the hidden file .openshift_state_install.json like all the assets and other data.

```mermaid
flowchart TD
    A[Check if openshift-config/agent-auth-token exists] -->|No| B[Create secret with 3 tokens and public key]
    A -->|Yes| C[Retrieve tokens from existing secret]
    B --> D[Generate new tokens if secret does not exist]
    C --> D[Check if token is older than 24 hours]
    D -->|Yes| E[Generate new tokens, update asset store, and secret]
    D -->|No| F[Retrieve token, update asset store, use token for ISO]
    F --> G[Node boots and checks token expiry]
    G -->|Expired| H[Display message: 'Authentication token expired. Generate a new ISO.']
    G -->|Not expired| I[Proceed with installation using retrieved token]
```
## Authorization

To implement authorization for ABI, a new security definition is added in the `swagger.yaml` file in AS. The new security definition is as follows:

```yaml
watcherAuth:
  type: apiKey
  in: header
  name: Watcher-Authorization
```

The required endpoints are annotated with this security definition. For example:

```yaml
/v2/clusters:
  get:
    tags:
      - installer
    security:
      - userAuth: [admin, read-only-admin, user]

>       - watcherAuth: []

Reviewer 2: For watcherAuth, what does an empty array ([]) mean? Shouldn't there be something like read-only listed?

Author: The empty array under watcherAuth indicates that no specific roles are required—this is intentional and specific to the ABI flow. Unlike the roles mentioned for userAuth, which is tied to RHSSO auth type and uses IAM roles like admin, user, etc., ABI doesn’t use IAM policies. Instead, access control is based on the endpoint behavior itself. 

For example, the watcherAuth persona (like the wait-for user) is only allowed to perform safe read-only actions like GET, while userAuth might have both GET and POST access, implying read-write. So the roles aren't declared because ABI evaluates access by persona and method, not by static role mapping.

    description: Retrieves the list of OpenShift clusters.
    operationId: v2ListClusters
```

### Authorization Handler

The Assisted Service API includes an authorization handler implemented in the `pkg/auth/agent_local_authz_handler.go` file. The `AgentLocalAuthzHandler`  is responsible for verifying that the `auth_scheme` in the token matches the required `authScheme` for the endpoint. The authorization logic works as follows:

1. The `auth_scheme` claim in the JWT token is extracted.
2. If the claim is malformed or missing, an error is returned.

> 3. The `auth_scheme` claim embedded in the token specifies the type of authorization (e.g., `watcherAuth`) and is compared with the `authScheme` expected by the endpoint. This ensures that the token matches the intended purpose of the request.  

Reviewer 2: Is /v2/clusters above the only endpoint that supports watcherAuth?

Author: No, the full details are in the swagger.yaml here https://github.com/openshift/assisted-service/blob/master/swagger.yaml#L169
